### PR TITLE
Add CI image to run tests on GitHub runner

### PR DESCRIPTION
**Pull request by crohr**
Allows using a self hosted GitHub Action runner to aggressively speed-up the tests. Only unit tests are passing for now, in about 10min (with cold cache, e.g. https://github.com/opf/openproject/pull/8858/checks?check_run_id=1459526934). I'd need some help understanding why the feature tests aren't passing.

I need your opinion on: do we merge with just the unit tests for now, so that we already get the perf boost for that part, or do we wait until the full suite is green?

### CODE REVIEW

File: spec/support/capybara.rb [hunk 1]
```diff
@@ -9,7 +9,7 @@
   Capybara.default_max_wait_time = 4
   Capybara.javascript_driver = :chrome_en
 
-  port = ENV.fetch('CAPYBARA_SERVER_PORT', '0').to_i
+  port = ENV.fetch('CAPYBARA_SERVER_PORT', ENV.fetch('TEST_ENV_NUMBER', '1').to_i + 3000).to_i
```
Comment (b12f) [inline spec/support/capybara.rb:12]: Can't you set `CAPYBARA_SERVER_PORT` directly to something like 3003?

File: script/ci/cache_prepare.sh [hunk 2]
```diff
@@ -38,9 +38,9 @@ run() {
   eval $2;
 }
 
-run "bundle exec rake db:migrate webdrivers:chromedriver:update webdrivers:geckodriver:update"
+run "bundle exec rake db:create db:migrate webdrivers:chromedriver:update webdrivers:geckodriver:update"
 
-run "for i in {1..3}; do (cd frontend; npm install && break || sleep 15;) done"
+run "cd frontend; npm install ; cd -"
```
Comment (b12f) [inline script/ci/cache_prepare.sh:43]: we could make `run` eval in a separate context (e.g. `(eval $2)`) that way we don't need to cd back after commands

File: docker/ci/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+COPY ./entrypoint.sh /usr/sbin/entrypoint.sh
+
+VOLUME [ "/usr/local/bundle", "/home/$USER/openproject", "/home/$USER/openproject/tmp" ]
+
+WORKDIR /home/$USER/openproject
+
+ENTRYPOINT [ "/usr/sbin/entrypoint.sh" ]
+
+EXPOSE 3000
```
Comment (b12f) [inline docker/ci/Dockerfile]: Is this necessary? we could set it in the docker command or the compose file anyway, and even expose 3312 if we wanted to

File: docker-compose.yml [hunk 4]
```diff
@@ -149,10 +150,39 @@ services:
...
+    depends_on:
+      - selenium-hub
+
+  # https://vitobotta.com/2019/09/04/rails-parallel-system-tests-selenium-docker/
   selenium-hub:
     image: selenium/hub:latest
     container_name: selenium-hub
     hostname: selenium-hub
+    environment:
+      GRID_MAX_SESSION: 16
```
Comment (b12f) [inline docker-compose.yml]: We could perhaps even set this in the env, allowing you to change it locally / on ci

File: docker-compose.yml [hunk 5]
```diff
@@ -168,13 +198,13 @@ services:
       - /dev/shm:/dev/shm
     networks:
       - testing
-    ports:
-      - 5900:5900
```
Comment (b12f) [inline docker-compose.yml:172]: This is probably so you can start multiple instances right? Being able to vnc into the container is really nice though. Could we maybe automatically let docker just set this to `5900 + $whateverNextPortIsFree`? Or alternatively, can grid start multiple browsers in one container, and then we only have vnc there?

But I'm okay sacrificing vnc for speed, so if it's too hard just leave it out.

File: spec/support/capybara.rb (hunk 1 above)
Comment (crohr) [reply spec/support/capybara.rb:12]: parallel_spec will start threads and the only thing differentiating those threads is the `TEST_ENV_NUMBER` env variable injected by parallel_spec. So we need to use this to dynamically adjust the port used in the feature tests (but I might be wrong).

File: script/ci/cache_prepare.sh (hunk 2 above)
Comment (crohr) [reply script/ci/cache_prepare.sh:43]: Sure, and we don't need that `run` prefix either. Not sure what it was for.

File: docker-compose.yml (hunk 4 above)
Comment (crohr) [reply docker-compose.yml]: Yes you're right it should use the same `CI_JOBS` var used to specify the number of parallel jobs.

File: docker-compose.yml (hunk 5 above)
Comment (crohr) [reply docker-compose.yml:172]: Yes the grid supports starting multiple sessions within the same container, although not recommended. I was indeed going for the `docker-compose scale chrome=XX`, but if we find that the multiple-session within one container works for us, even better.

File: docker/ci/Dockerfile (hunk 3 above)
Comment (crohr) [reply docker/ci/Dockerfile]: I didn't find a way to dynamically expose `CI_JOBS` ports from the docker-compose file. Hence that (ugly but limited to the CI dockerfile) solution. An alternative would be to generate an additional Compose file to override it, but that seemed overkill.

File: docker-compose.yml (hunk 5 above)
Comment (b12f) [reply docker-compose.yml:172]: I've never tried it and I'm not sure what exactly it is  used for, but here it shows you can set a range for the host ports (see the last example int he short syntax list)

https://docs.docker.com/compose/compose-file/#ports

Perhaps that would enable port-autoselecting behaviour

File: docker/ci/Dockerfile (hunk 3 above)
Comment (crohr) [reply docker/ci/Dockerfile]: Now that I think about it, dynamically appending a list of EXPOSE instructions at the end of this Dockerfile would also work fine.

File: spec/support/capybara.rb (hunk 1 above)
Comment (b12f) [reply spec/support/capybara.rb:12]: Okay, I would at least then write a comment there to explain why this check is so weird

**Comment by crohr** [comment]
Superseded by #8943. We won't use the selenium hub route since it causes a lot of issues with tests executing the frontend JS.

**Pull request closed without merging**